File: mobile/core/i18n/src/main/kotlin/af/velro/core/i18n/Numerals.kt
package af.velro.core.i18n

import af.velro.domain.Locale
import af.velro.domain.MoneyValue
import af.velro.domain.minorDigits
import java.math.BigDecimal

/**
 * Numerals and money, formatted for the reader.
 *
 * Storage is always Latin digits; this is a display concern only. Dari and
 * Pashto readers generally expect Eastern Arabic-Indic digits in prose and
 * Latin digits in tables of figures, so the choice is a preference with a
 * per-locale default rather than an assumption.
 */
object Numerals {

    private const val EASTERN = "۰۱۲۳۴۵۶۷۸۹"

    /** Set from the user's preference; defaults per locale. */
    @Volatile
    var useEasternDigits: Boolean? = null

    fun easternByDefault(locale: Locale): Boolean = locale != Locale.ENGLISH

    fun format(value: Number, locale: Locale): String =
        localise(value.toString(), locale)

    fun localise(text: String, locale: Locale): String {
        val eastern = useEasternDigits ?: easternByDefault(locale)
        if (!eastern) return text
        return buildString(text.length) {
            for (ch in text) {
                append(if (ch in '0'..'9') EASTERN[ch - '0'] else ch)
            }
        }
    }

    /** Always Latin digits: used for phone numbers, plates and codes. */
    fun latin(text: String): String = buildString(text.length) {
        for (ch in text) {
            val index = EASTERN.indexOf(ch)
            append(if (index >= 0) ('0' + index) else ch)
        }
    }
}

object MoneyFormatter {

    /**
     * "۵۰۰ افغانی" or "500 AFN".
     *
     * The amount is derived from integer minor units; no float touches this
     * path, so a fare never renders as 449.99999.
     */
    /**
     * The same rendering, from the parts a server sends.
     *
     * Error contexts and notification payloads carry `<name>_minor` plus a
     * `currency` rather than a MoneyValue, and Strings needs to turn those into
     * a sentence without the caller building a domain object first.
     */
    fun format(amountMinor: Long, currency: String, strings: Strings): String =
        format(MoneyValue(amountMinor, currency), strings)

    fun format(money: MoneyValue, strings: Strings): String {
        val digits = minorDigits(money.currency)
        val negative = money.amountMinor < 0
        // The sign is applied here, not left to BigDecimal, so that the digits
        // are grouped and localised without it and the sign can be attached to
        // an isolated run -- see [signed].
        val magnitude = if (negative) -money.amountMinor else money.amountMinor
        val major = BigDecimal(magnitude).movePointLeft(digits)
        val plain =
            if (major.stripTrailingZeros().scale() <= 0) major.toBigInteger().toString()
            else major.stripTrailingZeros().toPlainString()

        val grouped = group(plain)
        val symbol = strings["common.label.currency_afn"]
        val number = Numerals.localise(grouped, strings.locale)
        return "${signed(number, negative = negative)} $symbol"
    }

    /**
     * A number with a sign that stays on the correct side of it.
     *
     * A leading "-" in front of Arabic-Indic digits inside a right-to-left
     * paragraph does not stay put. The digits are bidi class AN and the sign
     * is neutral against them, so the algorithm resolves it to the paragraph
     * direction and moves it to the other end: the driver's ledger rendered
     * `-۵۵ افغانی` as `۵۵- افغانی`, with the sign sitting between the number
     * and the currency. On a screen that exists to tell a driver what he owes,
     * a minus that has drifted off its number is not a typographic complaint.
     *
     * The isolate is what pins it. LRI..PDI makes the sign and the digits one
     * left-to-right run, which is how a signed number is read in Dari and
     * Pashto as well -- numbers are read left to right in both.
     *
     * Only signed numbers are wrapped. A positive amount has nothing that can
     * drift, and wrapping it would put two invisible characters into every
     * fare in the product for no gain.
     *
     * U+2212 rather than the ASCII hyphen: this is a minus, and the two were
     * already being used interchangeably on the same row of the same screen.
     */
    fun signed(number: String, negative: Boolean, showPlus: Boolean = false): String {
        val sign = when {
            negative -> "\u2212"
            showPlus -> "+"
            else -> return number
        }
        return "\u2066" + sign + number + "\u2069"
    }

    /** Thousands separators, applied to the integer part only. */
    private fun group(plain: String): String {
        val negative = plain.startsWith("-")
        val body = plain.removePrefix("-")
        val (whole, fraction) = body.split(".").let {
            it[0] to it.getOrNull(1)
        }
        val grouped = whole.reversed().chunked(3).joinToString(",").reversed()
        return buildString {
            if (negative) append("-")
            append(grouped)
            if (fraction != null) append(".").append(fraction)
        }
    }
}

File: ios/VelroCore/Sources/VelroCore/Numerals.swift
/// Digits, formatted for the reader.
///
/// Storage is always Latin digits; this is display only. Dari and Pashto
/// readers expect Eastern Arabic-Indic digits in prose, English readers Latin.
/// The same rule as the Android app's `Numerals`, so a fare reads the same on
/// both phones.
public enum Numerals {
    private static let eastern = Array("۰۱۲۳۴۵۶۷۸۹")
    private static let arabic = Array("٠١٢٣٤٥٦٧٨٩")

    public static func localise(_ text: String, _ locale: AppLocale) -> String {
        guard locale != .english else { return text }
        return String(text.map { ch in
            if let digit = ch.wholeNumberValue, ch.isASCII { eastern[digit] } else { ch }
        })
    }

    public static func format(_ value: some BinaryInteger, _ locale: AppLocale) -> String {
        localise(String(value), locale)
    }

    /// Always Latin digits: phone numbers, codes and plates.
    ///
    /// A Persian keyboard types ۰-۹ and an Arabic one ٠-٩; both reach the
    /// server as 0-9, or the number is refused as malformed.
    public static func latin(_ text: String) -> String {
        String(text.map { ch in
            if let index = eastern.firstIndex(of: ch) ?? arabic.firstIndex(of: ch) {
                Character(String(index))
            } else {
                ch
            }
        })
    }
}

public enum MoneyFormatter {
    private static let minorDigits = ["AFN": 2, "USD": 2, "EUR": 2, "PKR": 2]

    /// "۵۰۰ افغانی" or "500 AFN", from integer minor units.
    ///
    /// No floating point touches this path, so a fare never renders as
    /// 449.99999. Grouping and the sign follow the Android formatter exactly.
    /// `showPlus` marks a gain -- "+۵۰ افغانی" beside an offer above the ask --
    /// with the sign held to its number the same way a minus is.
    public static func format(minor: Int64, currency: String, strings: Strings, showPlus: Bool = false) -> String {
        let digits = minorDigits[currency] ?? 2
        let negative = minor < 0
        let magnitude = negative ? -minor : minor
        var scale: Int64 = 1
        for _ in 0..<digits { scale *= 10 }
        let whole = magnitude / scale
        let fraction = magnitude % scale

        var plain = group(String(whole))
        if fraction != 0 {
            var tail = String(fraction)
            tail = String(repeating: "0", count: digits - tail.count) + tail
            while tail.hasSuffix("0") { tail.removeLast() }
            plain += "." + tail
        }
        let number = Numerals.localise(plain, strings.locale)
        return "\(signed(number, negative: negative, showPlus: showPlus && minor > 0)) \(strings["common.label.currency_afn"])"
    }

    /// A signed number whose sign stays beside it.
    ///
    /// Inside a right-to-left paragraph a leading minus drifts to the other
    /// end of Eastern digits. The isolate (LRI…PDI) pins sign and digits into
    /// one left-to-right run, which is how numbers are read in Dari and Pashto
    /// too. U+2212 because this is a minus, not a hyphen.
    public static func signed(_ number: String, negative: Bool, showPlus: Bool = false) -> String {
        let sign: String
        if negative {
            sign = "\u{2212}"
        } else if showPlus {
            sign = "+"
        } else {
            return number
        }
        return "\u{2066}" + sign + number + "\u{2069}"
    }

    private static func group(_ digits: String) -> String {
        var out = ""
        for (index, ch) in digits.reversed().enumerated() {
            if index > 0 && index % 3 == 0 { out.append(",") }
            out.append(ch)
        }
        return String(out.reversed())
    }
}
